barretenberg/cpp/src/barretenberg/stdlib/hash/poseidon2/poseidon2_permutation_classic.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+#include <array>
+#include <cstddef>
+#include <cstdint>
+
+#include "barretenberg/crypto/poseidon2/poseidon2_permutation.hpp"
+#include "barretenberg/stdlib/primitives/field/field.hpp"
+
+namespace bb::stdlib {
+
+using namespace bb;
+
+template <typename Params, typename Builder> class Poseidon2PermutationClassic {
+  public:
+    using NativePermutation = crypto::Poseidon2Permutation<Params>;
+    // t = sponge permutation size (in field elements)
+    // t = rate + capacity
+    // capacity = 1 field element (256 bits)
+    // rate = number of field elements that can be compressed per permutation
+    static constexpr size_t t = Params::t;
+    // d = degree of s-box polynomials. For a given field, `d` is the smallest element of `p` such that gdc(d, p - 1) =
+    // 1 (excluding 1) For bn254/grumpkin, d = 5
+    static constexpr size_t d = Params::d;
+    // sbox size = number of bits in p
+    static constexpr size_t sbox_size = Params::sbox_size;
+    // number of full sbox rounds
+    static constexpr size_t rounds_f = Params::rounds_f;
+    // number of partial sbox rounds
+    static constexpr size_t rounds_p = Params::rounds_p;
+    static constexpr size_t NUM_ROUNDS = Params::rounds_f + Params::rounds_p;
+
+    using FF = typename Params::FF;
+    using State = std::array<field_t<Builder>, t>;
+    using NativeState = std::array<FF, t>;
+
+    using RoundConstants = std::array<FF, t>;
+    using RoundConstantsContainer = std::array<RoundConstants, NUM_ROUNDS>;
+    static constexpr RoundConstantsContainer round_constants = Params::round_constants;
+    // static constexpr MatrixDiagonal internal_matrix_diagonal = Params::internal_matrix_diagonal;
+
+    /**
+     * @brief Circuit form of Poseidon2 permutation from https://eprint.iacr.org/2023/323.
+     * @details The permutation consists of one initial linear layer, then a set of external rounds, a set of internal
+     * rounds, and a set of external rounds.
+     * @param builder
+     * @param input
+     * @return State
+     */
+    static State permutation(Builder* builder, const State& input);
```

Comment:
why 2 functions? Unless I'm missing something, I would suggest merging them into one and using constrained template parameters (`requires IsGoblinBuilder<Builder>`)